_RELEASE/Packs/cube/Scripts/utils.lua

```diff
@@ -1,21 +1,552 @@
+	return r * math.cos(a), r * math.sin(a);
+end
+
+function getSign(number) 
+	--[[
+	Takes a number and returns it's sign value.
+	
+	Negatives return -1, positives return 1, and zero returns 0.
+	]]
+	if number < 0 then
+		return -1;
+	elseif number > 0 then
+		return 1;
+	end
+	return 0;
+end
+
+-- A data structure that stores results for prime numbers as a way to "cache" results.
+-- This helps make isPrime run faster by knocking out calculations that were already done.
+local prime_memoization = {
+	[0] = false,
+	[1] = false,
+	[2] = true
+};
+function isPrime(integer, divisor)
+	--[[
+	Determines if an integer is a prime number or not.
+	]]
+	-- Makes the divisor parameter optional. By default, it should start at 2.
+	divisor = divisor or 2; 
+
+	if (prime_memoization[integer] ~= nil) then
+    	return prime_memoization[integer];
+  	end
+	
+	if (integer % divisor == 0) then
+    	prime_memoization[integer] = false;
+		return false;
+	elseif (divisor * divisor > integer) then
+    	prime_memoization[integer] = true;
+		return true;
+	end
+	-- Make a recursive call to try other factors.
+	return isPrime(integer, divisor + 1); 
+end
+
+function inverseLerp(initial, final, value)
+	--[[
+	Inverse linear interpolation function. Takes two number values, initial
+	and final, and then a third number "value". Returns the percentage of
+	"value" between initial and final.
+	]]
+	return (value - initial) / (final - initial);
+end
+
+function lerp(initial, final, i) 
+	--[[
+	Linear interpolation function. Takes number value initial and returns a
+	value that is (i * 100) percent to final value.
+	
+	i is a number value between 0 and 1
+	
+	Thanks to Zly for showing me the "precise" method
+	]]
+	return (1 - i) * initial + i * final;
+end
+
+function printTable(luaTable, whitespace)
+	--[[
+	Prints a table to the console. This includes all key value pairs, and recursively prints out any 
+	tables of additional dimensions
+
+	Formatting of table contents are done with the "whitespace" character. Courtesy of Zly.
+	]]
+	whitespace = whitespace or "";
+	
+	if (type(luaTable) ~= "table") then return end;
+	
+	for k, v in pairs(luaTable) do
+		print(whitespace..tostring(k), v);
+		if (type(v) == "table" and v ~= luaTable) then
+			tablePrint(v, whitespace.."\t");
+		end
+	end
+end 
+
+function randomFloat(minimum, maximum)
+	--[[
+	Returns a random value between minimum and maximum, but is a floating point
+	number.
+	]]
+	return math.random() * (maximum - minimum) + minimum;
+end
+
+function roundFloat(floatingNumber)
+	--[[
+	Takes in a number value and follows true rounding rules.
+	Decimals < .5 will round down, while >= .5 will round up.
+	
+	Returns the rounded number value
+	]]
+	return math.floor(floatingNumber + 0.5);
+end
+
+function shuffle(t)
+	--[[
+		"Shuffles" an array by swapping elements randomly across a table.
+	]]
+	local j
+	for i = #t, 2, -1 do
+		j = math.random(i)
+		t[i], t[j] = t[j], t[i]
+	end
+	return t
+end
```

Comment:
Replace by this:
```lua
function shuffle(t)
  for i = 1, #t do
    local j = math.random(i)
    t[i], t[j] = t[j], t[i]
  end
end
```
Tables are passed as references and return is redundant. Starting from 1 to the length of the table for a proper shuffle and not "pointlessly restricted" area shuffle.

Reply:
it did a poor shuffle, more like a semi-shuffle lol.

Reply:
Well if I follow your change, good luck explaining to everyone how you broke their level pack by making shuffles no longer return.

Reply:
Are you an idiot or what? I just fucking said they are passed as **references**

Reply:
How about we have in-place and not-in-place variants then?
A proper not-in-place version could also be useful if you need two different permutations of an array, as it would effectively create a copy of it in a random order.

Reply:
Learn Lua ffs, I've been telling you this shit many god damn times.

Reply:
> 
> 
> Well if I follow your change, good luck explaining to everyone how you broke their level pack by making shuffles no longer return.

That wouldn't break existing packs as they will keep a copy of the old utils.lua file in there. This is just changing the new template, you're not changing files on everyone's workshops or whatever.

Reply:
Like okay i understand that default packs are using shuffle in a shit way aka:
```lua
keys = { 0, 0, 1, 1, 2, 2, 3, 4, 4, 5, 6, 7, 7, 7, 8, 9, 9 }
keys = shuffle(keys)
```
but let's damn fix that to:
```lua
keys = { 0, 0, 1, 1, 2, 2, 3, 4, 4, 5, 6, 7, 7, 7, 8, 9, 9 }
shuffle(keys)
```
Easy to do since it's only applied to default levels, not to existing ones from other people. :/

Reply:
Fine. It's just gonna be annoying to have to show how literally every level file is going to have changes.

Reply:
You don't have to show this..?

Reply:
I mean the pull request is gonna show every change I make. But whatever, it's fine.

Reply:
Eh, I don't see that as a problem.